src/io/dataset_loader.cpp

```diff
@@ -222,7 +240,14 @@ Dataset* DatasetLoader::LoadFromFile(const char* filename, int rank, int num_mac
       text_data.clear();
     } else {
       // sample data from file
-      auto sample_data = SampleTextDataFromFile(filename, dataset->metadata_, rank, num_machines, &num_global_data, &used_data_indices);
+      std::vector<std::string> sample_data;
+      std::vector<data_size_t> sampled_indices;
+      if (ctr_provider == nullptr) {
+        sample_data = SampleTextDataFromFile(filename, dataset->metadata_, rank, num_machines, &num_global_data, &used_data_indices);
+      } else {
+        sample_data = SampleTextDataFromFileWithIndices(filename, dataset->metadata_, rank,
+          num_machines, &num_global_data, &used_data_indices, &sampled_indices);
+      }
```

Comment:
move the `if (ctr_provider == nullptr) ... else ...` into SampleTextDataFromFile, then decide which function to call? to reduce duplicate codes.